ath79: Add support for TP-Link EC230-G1

izuannazrin opened the pull request:
Introduction: This device is sold to and distributed by ISPs (TIME in Malaysia). Hardware-wise, it is identical to Archer C6 (EU) but with 16MB flash. However software-wise it is very different (partition table, label MAC location, bootloader).

Specification:
- CPU: Qualcomm Atheros QCA9563-AL3A
- RAM: Zentel A3R1GE40JBF-8E (128MB)
- Flash: GigaDevice GD25Q127CSIG (16MB)
- 5GHz PHY: Qualcomm Atheros QCA9886
- Ethernet PHY: Qualcomm Atheros QCA8337N-AL3C (5 port 1000BASE-T + SGMII + ?)

Note: Some original firmware doesn't have the ability to update through the wizard (web interface or RESET button on boot), thus I cannot confirm flashing it that way for now, until somebody can tag along here.
For now, you'll need to use UART and tftp to install new firmware.

Prerequisite:
- Prepare USB TTL adapter, 3 jumper wire (test pad would be about size of wire diameter) and a kind, sentient being, also known as person.
- Set up TFTP server to serve the sysupgrade image. Refer OpenWrt wiki [generic.flashing.tftp](https://openwrt.org/docs/guide-user/installation/generic.flashing.tftp) to set it up.

Installation:
- Disassemble the casing. Two screw on the bottom then pry the top casing with brute force.
- Pry off metal shielding to expose the CPU circuitry.
- Connect network cable between PC and device.
- With help of other people, have them hold jumper wire over the tiny UART test pad. UART Tx would be near the small antenna labeled TP_UART_IN, and UART Rx would be near CPU labeled TP_UART_OUT.
- Open serial terminal of your liking and power on the device.
- Interrupt boot process when you see `Hit any key to stop autoboot:`
- On prompt `ath>`, enter following command, replacing <> where necessary:
```
	tftpboot 0x80020000 <your openwrt sysupgrade.bin>
	erase 0x9f020000 +0x<hex of size of sysupgrade.bin>
	cp.b 0x80020000 0x9f020000 0x<hex of size of sysupgrade.bin>
	bootm 0x9f020000
```
- Done. Let it boot.

Signed-off-by: Izuan Nazrin <zerro.tur@gmail.com>

GoliathLabs commented:
What about the firmware upgrade section on the router's web interface? Have you tried to rename your firmware file to something else (for example, to a file name that the ISP uses)?

GoliathLabs commented:
But thumbs up for your effort to get OpenWrt working on this device 😄.

izuannazrin commented:
> What about the firmware upgrade section on the router's web interface? Have you tried to rename your firmware file to something else (for example, to a file name that the ISP uses)?

Nope, there are no firmware upgrade page in the web interface unlike in other TP-Link routers I've seen. Probably to avoid end-user from upgrading to custom firmware, and/or they can upgrade it through ACS (I've seen attempt to acs.time.com.my in the serial output).

Even typical tftp button trigger didn't work, like pushing Reset button upon power on.


>But thumbs up for your effort to get OpenWrt working on this device 😄.

Thanks XD!

GoliathLabs commented:
Okay, strange, I did a quick search and a guy did a sysupgrade on his EC230-G1 in the video (time ISP). I can't understand Malay so this could be a fake or something. (https://www.youtube.com/watch?v=oZ0LAzmNua4)

Hopefully this one will be merged 👍🏼

izuannazrin commented:
> Okay, strange, I did a quick search and a guy did a sysupgrade on his EC230-G1 in the video (time ISP). I can't understand Malay so this could be a fake or something. (https://www.youtube.com/watch?v=oZ0LAzmNua4)

Oh damn I didn't search for it online. Trust me, mine didn't have it there (the place I expected it to be). I guess his firmware version is different than mine, and I got the bad (bad as in bad kitty) one :(

Maybe sometime in the future I can confirm this, or anybody else?

GoliathLabs commented:
Yes, no problem. I checked to be on the safe side, because I want everyone to have the option to use a different firmware than the one from the manufacturer. Opening the device, connecting a TTL adapter and doing some operations in the serial console to get the OpenWrt firmware on the device will surely only get a few people to use it. That's why I was looking for another method.

adschm commented on target/linux/ath79/image/generic-tp-link.mk:
```diff
@@ -426,6 +426,16 @@ define Device/tplink_eap245-v3
+  SOC := qca9563
+  IMAGE_SIZE := 16000k
+  DEVICE_MODEL := EC230-G1
+  TPLINK_BOARD_ID := EC230-G1
```
That's interesting. I thought TPLINK_BOARD_ID is only required for tp-link-safeloader, which we don't need here since we can't flash? So, can this be removed, or is it needed anywhere else I don't think of?

izuannazrin replied:
You're right, It's not needed. Sorry I referenced it from other board definition, and didn't know what those variables meant.

adschm commented on target/linux/ath79/image/generic-tp-link.mk:
```diff
@@ -426,6 +426,16 @@ define Device/tplink_eap245-v3
+  IMAGE_SIZE := 16000k
+  DEVICE_MODEL := EC230-G1
+  TPLINK_BOARD_ID := EC230-G1
+  DEVICE_PACKAGES := kmod-ath10k-ct ath10k-firmware-qca9888-ct
```
So, looking at this I wanted to check how much RAM the device has. But the commit message doesn't tell. Please make the specifications more complete: RAM size, Flash size, ...

izuannazrin committed:
ath79: Add support for TP-Link EC230-G1

Introduction: This device is sold to and distributed by ISPs (TIME in
Malaysia). Hardware-wise, it is identical to Archer C6 (EU) but with
16MB flash. However software-wise it is very different (partition
table, label MAC location, bootloader).

Specification:
	CPU: Qualcomm Atheros QCA9563-AL3A
	RAM: Zentel A3R1GE40JBF-8E (128MB)
	Flash: GigaDevice GD25Q127CSIG (16MB)
	5GHz PHY: Qualcomm Atheros QCA9886
	Ethernet PHY: Qualcomm Atheros QCA8337N-AL3C (5 port 1000BASE-T
	              + SGMII + ?)

Note: Some original firmware doesn't have the ability to update through
the wizard (web interface or RESET button on boot), thus I cannot
confirm flashing it that way for now, until somebody can tag along here.
For now, you'll need to use UART and tftp to install new firmware.

Prerequisite:
- Prepare USB TTL adapter, 3 jumper wire (test pad would be about size
  of wire diameter) and a kind, sentient being, also known as person.
- Set up TFTP server to serve the sysupgrade image. Refer OpenWrt wiki
  generic.flashing.tftp to set it up.

Installation:
- Disassemble the casing. Two screw on the bottom then pry the top casing
  with brute force.
- Pry off metal shielding to expose the CPU circuitry.
- Connect network cable between PC and device.
- With help of other people, have them hold jumper wire over the tiny
  UART test pad. UART Tx would be near the small antenna labeled TP_UART_IN,
  and UART Rx would be near CPU labeled TP_UART_OUT.
- Open serial terminal of your liking and power on the device.
- Interrupt boot process when you see "Hit any key to stop autoboot:"
- On prompt "ath>", enter following command, replacing <> where
  necessary:
	tftpboot 0x80020000 <your openwrt sysupgrade.bin>
	erase 0x9f020000 +0x<hex of size of sysupgrade.bin>
	cp.b 0x80020000 0x9f020000 0x<hex of size of sysupgrade.bin>
	bootm 0x9f020000
- Done. Let it boot.

Signed-off-by: Izuan Nazrin <zerro.tur@gmail.com>

izuannazrin commented:
Hi, chiming in back after a long time. Is there anything else I should do to push this?

izuannazrin commented:
@adschm sorry to ping you. Should I close this first and re-'pull request'(?) in the future? After all, there's already too many commit that made manual merging required.

adschm commented:
> @adschm sorry to ping you. Should I close this first and re-'pull request'(?) in the future? After all, there's already too many commit that made manual merging required.

https://openwrt.org/docs/guide-developer/frequent-pr-mistakes#do_not_multi-post_or_re-post_prs

adschm commented on target/linux/ath79/image/generic-tp-link.mk:
```diff
@@ -426,6 +426,18 @@ define Device/tplink_eap245-v3
+  DEVICE_PACKAGES := kmod-ath10k-ct ath10k-firmware-qca9888-ct
+  TPLINK_HWID := 0x4E8C8701
+  TPLINK_HWREV := 0x00060034
+  TPLINK_FLASHLAYOUT := 16Mqca
```
Where do these three come from?

izuannazrin replied:
From the original image header, parsed [here (gist)](https://gist.github.com/izuannazrin/0c32f269b7ff008db04ad0403a22acc0). Though from my testing it's not necessary but for the sake of correctness, \*insert idk emoji\* why not

izuannazrin commented:
Updated to rebase to master and fix merge conflict.

drabikp commented:
@GoliathLabs @izuannazrin Hi guys, I just wanted to let you know that as far as I understand, there is no good and bad version of a firmware. When the router is powered on for the very first time (factory configuration), you can access High Level Administration web using 192.168.0.1/superadmin (or whaterver the IP would be). The web contains section Advanced -> System Tools -> Administration, where you can select which functions will be enabled for the user - and one of them is Firmware Update. This superadmin interface works only once and you need to perform a specific procedure to achieve "real" factory reset.

Here are the steps to make superadmin interface available:
1. Power on the router and connect it to the PC via LAN port.
2. telnet 192.168.0.1 (or whatever the default IP is) *within 10 minutes of the device being powered on*
3. Enter "reset" as password; you should get message like “You can reset to TP-Link Config now."
4. Press the reset button on the router for at least 5 seconds and after reboot you are good to go - superadmin interface should be available.

This is exempt from the attached PDF I found after some googling.
That being said, I tried to build an image from your fork, but uploading it to router via web interface failed. I tried to rename the image to EC230-G1(EU)v1_2.0.0_0.9_v6034.0_Build_210226_Rel.28368n.bin (I saw some firmware online...) but it didn't work.

Do you guys have any further ideas how to deal with this?
Cheers!

[Agile Config Suite 2.0 Operation Manual.pdf](https://github.com/openwrt/openwrt/files/9616985/Agile.Config.Suite.2.0.Operation.Manual.pdf)

avatar7008 commented:
I have access to a unit with admin access. Would you like to test it? @izuannazrin

itshog commented:
@izuannazrin Sorry to bother you, just wanted to ask if this still works. I have this router and the original firmware is extremely buggy.

hauke commented on target/linux/ath79/dts/qca9563_tplink_ec230-g1.dts:
```diff
@@ -0,0 +1,170 @@
+		};
+
+		wlan2g {
+			label = "green:wlan2g";
```
Please us this instead:
```suggestion
			color = <LED_COLOR_ID_GREEN>;
			function = LED_FUNCTION_WLAN_2GHZ;
```

hauke commented on target/linux/ath79/dts/qca9563_tplink_ec230-g1.dts:
```diff
@@ -0,0 +1,170 @@
+		};
+
+		wlan5g {
+			label = "green:wlan5g";
```
Please us this instead:
```suggestion
			color = <LED_COLOR_ID_GREEN>;
			function = LED_FUNCTION_WLAN_5GHZ;
```

hauke commented on target/linux/ath79/dts/qca9563_tplink_ec230-g1.dts:
```diff
@@ -0,0 +1,170 @@
+		};
+
+		lan {
+			label = "green:lan";
```
Please us this instead:
```suggestion
			color = <LED_COLOR_ID_GREEN>;
			function = LED_FUNCTION_LAN;
```

hauke commented on target/linux/ath79/dts/qca9563_tplink_ec230-g1.dts:
```diff
@@ -0,0 +1,170 @@
+
+		wan {
+			label = "green:wan";
+			gpios = <&gpio 7 GPIO_ACTIVE_LOW>;
```
Please us this instead:
```suggestion
			color = <LED_COLOR_ID_GREEN>;
			function = LED_FUNCTION_WAN;
```

hauke commented on target/linux/ath79/dts/qca9563_tplink_ec230-g1.dts:
```diff
@@ -0,0 +1,170 @@
+		};
+
+		wan {
+			label = "green:wan";
```
Please us this instead:
```suggestion
			color = <LED_COLOR_ID_GREEN>;
			function = LED_FUNCTION_LAN;
```

hauke commented on target/linux/ath79/dts/qca9563_tplink_ec230-g1.dts:
```diff
@@ -0,0 +1,170 @@
+		};
+
+		wan_fail {
+			label = "amber:wan";
```
Please us this instead:
```suggestion
			color = <LED_COLOR_ID_AMBER>;
			function = LED_FUNCTION_WAN;
```

hauke commented:
Please rebase this PR.

madIlama commented:
Hi, i have EC230-G1 with admin/upgrade page open. After cherrypick to compile using latest 25.12, it seems cant uploaded to OEM upgrade page. Even C6 seems have this "magic number" problem too : https://openwrt.org/toh/tp-link/archer_c6_v2#flash_using_web_interface

but in C6 there is failsafe TFTP boot and on EC230, seems nada : https://community.tp-link.com/en/home/forum/topic/251448

So unless we got header right ( i  still use @izuannazrin header intact) serial still only way ( i still not tested, preparing tools)
